ViewEnvironmentUI/Sources/ViewEnvironmentPropagating.swift

```diff
@@ -105,22 +126,248 @@ extension ViewEnvironmentPropagating {
+    ///    assertion.
+    ///
+    @_spi(ViewEnvironmentWiring)
+    public var environmentDescendantsOverride: EnvironmentDescendantsProvider? {
+        get {
+            objc_getAssociatedObject(self, &AssociatedKeys.descendantsOverride) as? EnvironmentDescendantsProvider
+        }
+        set {
+            assert(
+                newValue == nil
+                    || environmentDescendantsOverride == nil,
+                "Attempting to set environment descendants override when one is already set."
+            )
+            objc_setAssociatedObject(self, &AssociatedKeys.descendantsOverride, newValue, .OBJC_ASSOCIATION_RETAIN)
+        }
+    }
+}
+
+/// A closure that is called when the `ViewEnvironment` needs to be updated.
+///
+public typealias ViewEnvironmentUpdateObservation = (ViewEnvironment) -> Void
+
+/// Describes the lifetime of a `ViewEnvironment` update observation.
+///
+/// The observation will be removed when `remove()` is called or the lifetime token is
+/// de-initialized.
+///
+/// ## SeeAlso ##
+/// - `addEnvironmentNeedsUpdateObserver(_:)`
+///
+public final class ViewEnvironmentUpdateObservationLifetime {
+    /// Removes the observation.
+    ///
+    /// This is called in `deinit`.
+    ///
+    public func remove() {
+        onRemove()
+    }
+
+    var onRemove: () -> Void
+
+    init(onRemove: @escaping () -> Void) {
+        self.onRemove = onRemove
+    }
+
+    deinit {
+        remove()
+    }
+}
+
+private enum ViewEnvironmentPropagatingNSObjectAssociatedKeys {
+    static var needsEnvironmentUpdate = NSObject()
+    static var needsUpdateObservers = NSObject()
+    static var ancestorOverride = NSObject()
+    static var descendantsOverride = NSObject()
+}
+
+extension ViewEnvironmentPropagating {
+    private typealias AssociatedKeys = ViewEnvironmentPropagatingNSObjectAssociatedKeys
+
+    public func setNeedsEnvironmentUpdate() {
+        needsEnvironmentUpdate = true
+
+        if !needsUpdateObservers.isEmpty {
+            let environment = environment
+
+            for observer in needsUpdateObservers.values {
+                observer(environment)
             }
+        }
+
+        if let observer = self as? ViewEnvironmentObserving {
+            observer.environmentDidChange()
 
-            descendant.setNeedsEnvironmentUpdate()
+            setNeedsApplyEnvironment()
+        }
+
+        setNeedsEnvironmentUpdateOnAppropriateDescendants()
+    }
+
+    private var needsUpdateObservers: [NSObject: ViewEnvironmentUpdateObservation] {
+        get {
+            objc_getAssociatedObject(
+                self,
+                &AssociatedKeys.needsUpdateObservers
+            ) as? [NSObject: ViewEnvironmentUpdateObservation] ?? [:]
+        }
+        set {
+            objc_setAssociatedObject(
+                self,
+                &AssociatedKeys.needsUpdateObservers,
+                newValue,
+                .OBJC_ASSOCIATION_RETAIN
```

Comment:
since this is UI code, should the nonatomic association be used? i assume this should never run concurrently

Reply:
I think you're right that this could/should be nonatomic! We should only be accessing this from the main actor/thread.

97b49ca

cc @watt

Reply:
Seems OK to me! I don't recall any reason it was done this way originally.